### ISSUE TITLE
Issue with pivot table

### ISSUE THREAD

**Issue by lokendrajoshi54** [radmen/adonis-lucid-soft-deletes#4]
While fetching relations facing issue with pivot table. When I use "whereHas" on model it does not look deleted_at in pivot table. Please check below example:

Route.query().whereHas('regions', (builder) => {
          builder.where('regions.id', queryParams.ward_id)
        })
there is pivot table "route_regions" which holds "route_id" and "region_id" and for some rows in  "route_regions" table there are some entries which have not null value for deleted_at. But above code fetching all records (including relations for which mappings are soft deleted).
I have to add condition for this in every query (Please check below):
Route.query().whereHas('regions', (builder) => {
          builder.where('regions.id', queryParams.ward_id)
          builder.whereNull('route_regions.deleted_at')
        })
Is there anything I missed or Is there any workaround for this problem?

**Comment by radmen** [comment]
Hey, thanks for feedback!

It's a longshot, but have you tried to use [pivotModel](https://adonisjs.com/docs/4.0/relationships#_pivotmodel)? You should create one and attach to it `@provider:Lucid/SoftDeletes` trait.

**Comment by lokendrajoshi54** [comment]
Hi, Thanks for reply,

I tried following but didn't work:
 Model.addGlobalScope(
      query => {
        query.whereNull(`${Model.table}.deleted_at`)
        if (query._PivotModel && query.relatedQuery) {
          query.whereNull(`${query._PivotModel.table}.deleted_at`);
        }
      },
      'soft_deletes'
    )

**Comment by radmen** [comment]
I'm not sure what you're trying to achieve. From the given code I don't see that you're using `SoftDeletes` trait. Also I'm not able to determine if you created a pivot model.

Could you give a more complete example of your models? Something that would allow me to recreate the problem? Thanks.

**Comment by lokendrajoshi54** [comment]
Sorry for late reply. Please check example below in my case:
Model:

`const Model = use('Model')

class GpsDevice extends Model {
  static boot () {
    super.boot()
    this.addTrait('@provider:Lucid/SoftDeletes')
  }

  vehicles() {
    return this.belongsToMany('App/Models/Vehicle').pivotModel('App/Models/VehicleGpsdevice')
  }

}`

Query:
GpsDevice.query().whereHas('vehicles', (builder) => {
builder.where('vehicle_type_id', 2)
}).with(vehicles).fetch()

The above query fetching records with deleted mappings in vehiclegpsdevice model. It seems its not checking deleted_at on pivot table. Can you please help me in that. Thanks

**Comment by radmen** [comment]
@lokendrajoshi54 thanks for the code example.  Could you also post contents of `VehicleGpsdevice` model (at least `boot()` method of that model)? Thanks!

**Comment by lokendrajoshi54** [comment]
Sure @radmen here is content of VehicleGpsdevice Model:
const Model = use('Model')

class VehicleGpsdevice extends Model {
    static boot () {
        super.boot()
        this.addTrait('@provider:Lucid/SoftDeletes')
      }
..
..
}

module.exports = VehicleGpsdevice

Same issue I faced with withTrashed() method, I'm trying to figure out if I missed anything, if you can tell me it would be helpful.
Thanks in advance

**Comment by radmen** [comment]
Thanks for the code. I'll try to check it tomorrow.

**Comment by radmen** [comment]
@lokendrajoshi54 I did some testing with the examples you gave.

`SoftDeletes` trait works with pivot model when fetching related records. Here's the test case I used:

https://github.com/radmen/adonis-lucid-soft-deletes/blob/5835dd4311dd174d10873afa3fca8786440025b0/test/softDeletes.spec.js#L314-L336

There's a problem with applying `SoftDeletes` global scope in `whereHas()`. Test case with example:

https://github.com/radmen/adonis-lucid-soft-deletes/blob/5835dd4311dd174d10873afa3fca8786440025b0/test/softDeletes.spec.js#L339-L359

So, TLDR:

* trait should work without any problem when fetching related items
* trait fails to apply its scope in `whereHas()` query; most likely it's a bug in Lucid

I'll try to verify if issues with `whereHas()` are indeed Lucid-related.

**Comment by lokendrajoshi54** [comment]
Hello @radmen 

Thanks for your reply and looking into issue. Same thing is applicable for withTrashed() method.

Is there any way I can help in this?

**Comment by radmen** [comment]
> Same thing is applicable for withTrashed() method.

In which context does this apply? `whereHas()`?

> Is there any way I can help in this?

First thing we need to determine is whether this is an issue with the package or Lucid.  If you think that the problem is with the package you could try writing a test case which fails - this would help me finding the bug,j

**Comment by radmen** [comment]
I was able to reproduce the problem on clean Adonis application. Its been reported: https://github.com/adonisjs/adonis-lucid/issues/429

**Comment by lokendrajoshi54** [comment]
@radmen Thanks for looking into this.

**Comment by lokendrajoshi54** [comment]
@radmen withTrashed() method is not working inside with method. Here is the snippet: 
let vehicles = await Vehicle.query().whereHas('gps_devices').with('gps_devices', (build) => {build.withTrashed()}).fetch()

**Comment by radmen** [comment]
@lokendrajoshi54 I'm aware of that. In fact it seems that this is something that is not possible in Lucid.

I've [asked](https://github.com/adonisjs/adonis-lucid/issues/429#issuecomment-471912431) about it, yet without any answer.
